Add support for SamsungTV encrypted models

epenet opened the pull request:
Add support for SamsungTV encrypted models

It can be tested using the following steps:

- create a new `custom_components\samsungtv` folder in your HA config folder
- copy all the files (including the translations) into the new folder. They can be downloaded from this link: https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/epenet/home-assistant-core/tree/samsungtv-encrypted/homeassistant/components/samsungtv
- edit `custom_components\samsungtv\manifest.json` to add a new `version` key: 
```json
{
  "domain": "samsungtv",
  "name": "Samsung Smart TV",
  "version": "2022.4.0-20220316",
  "documentation": "https://www.home-assistant.io/integrations/samsungtv",
  "...",
}
```
- adjust your configuration file for logging:
```yaml
logger:
    default: warning
    logs:
        homeassistant.components.samsungtv: debug
        custom_components.samsungtv: debug
        samsungtvws: debug
        samsungctl: debug
```

![image](https://user-images.githubusercontent.com/6771947/156597687-acca540f-5ccf-458e-b3e0-92c480384f8f.png)

epenet commented:
@sermayoral since I don't actually own an encrypted TV, I would very much appreciate if you could try out the example script and validate that:
- the "authenticator" works correctly to get a token and session_id
- the "remote" works correctly to send a key to the TV

Then we can tidy it up and move it to a separate library (it is compulsory to have a separate library for HA Core)

ooohaaa commented:
Hey @epenet, I followed your invitation, but I don't know how I can assist here. Guess I need some extra explanation. Thanks!

epenet commented:
You can test with the same as the original method, just with an updated download path.

- create a new `custom_components\samsungtv` folder in your HA config folder
- copy all the files (including the translations) into the new folder. They can be downloaded from this link: https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/epenet/home-assistant-core/tree/samsungtv-encrypted/homeassistant/components/samsungtv
- edit `custom_components\samsungtv\manifest.json` to add a new `version` key: 
```json
{
  "domain": "samsungtv",
  "name": "Samsung Smart TV",
  "version": "2022.4.0-20220308",
  "documentation": "https://www.home-assistant.io/integrations/samsungtv",
  "...",
}
```
- adjust your configuration file for logging:
```yaml
logger:
    default: warning
    logs:
        homeassistant.components.samsungtv: debug
        custom_components.samsungtv: debug
        samsungtvws: debug
        samsungctl: debug
```

![image](https://user-images.githubusercontent.com/6771947/156597687-acca540f-5ccf-458e-b3e0-92c480384f8f.png)

And there will be an added `encrypted` subfolder...

ooohaaa commented:
Thanks for your explanation - here is the log:
[HA-Core Log.txt](https://github.com/epenet/home-assistant-core/files/8227289/HA-Core.Log.txt)

sermayoral commented:
@epenet sorry for testing this so late, i had a very busy day :-(

I'm trying to test the authenticator script now, but i get the following error:

```
$ python authenticator_example.py
Traceback (most recent call last):
  File "authenticator_example.py", line 8, in <module>
    from .authenticator import SamsungTVEncryptedWSAsyncAuthenticator
ModuleNotFoundError: No module named '__main__.authenticator'; '__main__' is not a package
```

Edit: Sorry, i found the problem

sermayoral commented:
**authenticator_example**

port 8080

```
$ python authenticator_example.py
Please enter pin from tv: 2257
AES key: 7d41e3a531b6613e764e16dd68d588b2
AES encrypted: a856e27a0f98522b4868a90a4eee28453675fdc952f6527af5d977e7ec68c2b9792cad5e489a7d49bce522891e606a545c22580ddb31d2b37da5a52fdb724d49b3176b20de29f9c2772070edce83e2a784c1c756b523feea567178bc073ae615b50ac3277c2e58c13b1e1237737260d90f6028c0f16537e8e01fcf5cb57fa8f0
AES swapped: 6516b9effd3d5b0823f4222a2c59b9d813740c9987fb5bb16caed0cb1b32e4016253cf2b067965a7165ba7b2c0adacc5c11163cad9a7b8efc6a42f571aa0ffd167f22d3586a02f163863c2f664cad6384a8ac0fac1291bb57094a3da16a4736d8d9097a532729f20b08c1ef1ba3e521eb4b53c0e034453d3ab28015f4c3c2b9d
data buffer: 000000063635343332316516B9EFFD3D5B0823F4222A2C59B9D813740C9987FB5BB16CAED0CB1B32E4016253CF2B067965A7165BA7B2C0ADACC5C11163CAD9A7B8EFC6A42F571AA0FFD167F22D3586A02F163863C2F664CAD6384A8AC0FAC1291BB57094A3DA16A4736D8D9097A532729F20B08C1EF1BA3E521EB4B53C0E034453D3AB28015F4C3C2B9D
hash: 975e65af19c9fbffc41a971a3714a84a9381a46b
thirdLen: 138
hello: 010100000000000000009e00000006363534333231f824573835499faf5dc2c9cc12ed63b00c54bac8076733bfdc7d866f30cb2ebe23d9c51b71695c395e80ca8fb7b8a252e0fab3f5819104830bf938a70bb971bf1f93ea0e619d774b4ffbe948c9c4442cbfe5440640173043daf40bbe19697856241e65b0dadd715286511065472ba91294a7f480d395f6928f02a0e6633fe6e56468bb780c77971deee7a2033744f922b0e7935b0000000000
dest: 00000006363534333231f824573835499faf5dc2c9cc12ed63b00c54bac8076733bfdc7d866f30cb2ebe23d9c51b71695c395e80ca8fb7b8a252e0fab3f5819104830bf938a70bb971bf1f93ea0e619d774b4ffbe948c9c4442cbfe5440640173043daf40bbe19697856241e65b0dadd715286511065472ba91294a7f480d395f6928f02a0e6633fe6e5975e65af19c9fbffc41a971a3714a84a9381a46b
userId: 654321
pEncWBGx: f824573835499faf5dc2c9cc12ed63b00c54bac8076733bfdc7d866f30cb2ebe23d9c51b71695c395e80ca8fb7b8a252e0fab3f5819104830bf938a70bb971bf1f93ea0e619d774b4ffbe948c9c4442cbfe5440640173043daf40bbe19697856241e65b0dadd715286511065472ba91294a7f480d395f6928f02a0e6633fe6e5
pEncGx: 4fdf7cb88b8e28723cf08c4773ca492b3827073a1a8f95fd51326b9e2da6aed97007216ffaa62ba39f79a7b736ea3dc5b96d3d0f6552dd00692bd723eacc95b4c3d8ebb40618b01b53318e677f83a09579c82181e482e697461991cad43e926aca0cf3f02519f5c7cc7c15994c1bfa223706c0dcd7dcf8496c04bfa3b7b2c969
pGx: 2b72424c728b4413036427f897d9e864f088a796e278f15a304f7235e842695a41b77cc19cdd0fc09d6e371d8a765cb294b2f07c33ace40b86a7df3fb58880f58d48f276d7ce7b69d2f499db28946f9022a87a591d180ed8fbf335bf0cc6bb6a10b51a51b714dbb710c1a09d54afca8bce9189a367fa083c81ee5badd883c3cf
secret: 7e6db25792ee5b014b611f84c1f42f3378e7be8f34643c28e9a7ce31b4b8d985ad9fec6641eb6475ba921924ef163886191a3c677b9cacff18139c990d5440ffcee959298df801d54591f0d05eecb4a0947018390884833946feaf1827d4e065c8273d395a9376f758e0a449e63384efd50823c9fd0668d48108b2ae2086db1b
hash2: 6468bb780c77971deee7a2033744f922b0e7935b
secret2: 3635343332317e6db25792ee5b014b611f84c1f42f3378e7be8f34643c28e9a7ce31b4b8d985ad9fec6641eb6475ba921924ef163886191a3c677b9cacff18139c990d5440ffcee959298df801d54591f0d05eecb4a0947018390884833946feaf1827d4e065c8273d395a9376f758e0a449e63384efd50823c9fd0668d48108b2ae2086db1b
hash3: 6468bb780c77971deee7a2033744f922b0e7935b
Pin OK :)

dest_hash: 44ecb77f55277b4fdbfa61bd686e47dce9a9acdd
SKPrime: f58614a4d87a2d7fd603d4890ceae617817b1e6d
SKPrimeHash: f65f32ad9ea2ffa25da5386ce74a05fe58bd3317
Traceback (most recent call last):
  File "authenticator_example.py", line 42, in <module>
    loop.run_until_complete(main())
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\asyncio\base_events.py", line 568, in run_until_complete
    return future.result()
  File "authenticator_example.py", line 37, in main
    token, session_id = await _get_token(host, web_session, port)
  File "authenticator_example.py", line 24, in _get_token
    token = await authenticator.try_pin(pin)
  File "D:\T151602\Desktop\home-assistant-core\homeassistant\components\samsungtv\encrypted\authenticator.py", line 270, in try_pin
    self._sk_prime = result["sk_prime"]
KeyError: 'sk_prime'
```
typo in authenticator.py. Key defined in line 155 ("SKPrime") doesn't match with key defined in line 270 ("sk_prime")

**remote_example.py**

```
$ python remote_example.py
Traceback (most recent call last):
  File "remote_example.py", line 8, in <module>
    from remote import SamsungTVEncryptedWSAsyncRemote, SendRemoteKey
  File "D:\T151602\Desktop\home-assistant-core\homeassistant\components\samsungtv\encrypted\remote.py", line 18, in <module>
    from command import SamsungTVEncryptedCommand
  File "D:\T151602\Desktop\home-assistant-core\homeassistant\components\samsungtv\encrypted\command.py", line 8, in <module>
    class SamsungTVEncryptedCommand:
  File "D:\T151602\Desktop\home-assistant-core\homeassistant\components\samsungtv\encrypted\command.py", line 9, in SamsungTVEncryptedCommand
    def __init__(self, method: str, body: dict[str, Any]) -> None:
TypeError: 'type' object is not subscriptable
```

epenet commented:
I have adjusted the default port for the authenticator (`8000` => `8080`)
I have adjusted the authenticator for the KeyError (`sk_prime` => `SKPrime`)
I have adjusted the command for `object is not subscriptable` (added `from __future__ import annotations`) - I assume you are not on Python 3.9...

sermayoral commented:
@epenet i'm using python 3.7

**authenticator_example**

It works :-)

```
$ python authenticator_example.py
Please enter pin from tv: 0997
AES key: a81d67ca0535a2be13d5e3c51e52b3b4
AES encrypted: 9ab13b3e834f77a06abf4431f439b2242341a02dc594cde92b6fdabbfe7353b858ec153ae28509367132989e6074092ab2cd83a43e39e2c7002abf34c70e2b9ae062c3e65f5bd28b98f071ad55739c2386194ed2bd732ab3aa9920532bd8a8f286c3fcf8b74c7491e395789226564c3c467c1bb5a011d23c711192a5d304c1d0
AES swapped: 7caf9cbdc06b666d23ebca615e0666feb2b807091bf507404ddd18329cd64a91e513dc704298cce49c4c5656c42141a696354a7145127bcd94cdd2b0d632d87e332437f86ebe5a50a1512f3f54c71b791a88ecbaf562fbabe2731f27d851a764ca114dbe2c2c965df151cfc7401920faa04636b356b97dbe1da3a090004f8183
data buffer: 000000063635343332317CAF9CBDC06B666D23EBCA615E0666FEB2B807091BF507404DDD18329CD64A91E513DC704298CCE49C4C5656C42141A696354A7145127BCD94CDD2B0D632D87E332437F86EBE5A50A1512F3F54C71B791A88ECBAF562FBABE2731F27D851A764CA114DBE2C2C965DF151CFC7401920FAA04636B356B97DBE1DA3A090004F8183
hash: c729bf87cbcd9610bbd0948b43ae9c433a2791f0
thirdLen: 138
hello: 010100000000000000009e00000006363534333231f596d0966d38bdf42546fb2a06ae96161680381fbca62498e82903c36da100eba0c148cc1545db8f976a14423d95df7cac081b3722c2720c7ecc8d746d269319d309d36e432a1e32fea28dd7492692a71c7bf531d11a8f45ebb2a2834bb21e02e83ac7978396c03cfdd53256df124c09fdcae1711a9aeceaa83f3b8d8b2e70dcfe709b3e807dcaa9a9787f6a2f64475e9a70c1d80000000000
dest: 00000006363534333231f596d0966d38bdf42546fb2a06ae96161680381fbca62498e82903c36da100eba0c148cc1545db8f976a14423d95df7cac081b3722c2720c7ecc8d746d269319d309d36e432a1e32fea28dd7492692a71c7bf531d11a8f45ebb2a2834bb21e02e83ac7978396c03cfdd53256df124c09fdcae1711a9aeceaa83f3b8d8b2e70dcc729bf87cbcd9610bbd0948b43ae9c433a2791f0
userId: 654321
pEncWBGx: f596d0966d38bdf42546fb2a06ae96161680381fbca62498e82903c36da100eba0c148cc1545db8f976a14423d95df7cac081b3722c2720c7ecc8d746d269319d309d36e432a1e32fea28dd7492692a71c7bf531d11a8f45ebb2a2834bb21e02e83ac7978396c03cfdd53256df124c09fdcae1711a9aeceaa83f3b8d8b2e70dc
pEncGx: 0d9390dc45452716fb0ef5fa48d2537bb6712e1b05261aa185b05675d80e10a5c927d553051cba548b8ab343ed7d6c6494eb91e76f27f4dafc7fd196040706f3606bbd8664e183db489225c2f1ff783bf1e29a21f150460e1d1ff3301b54f0bad3f018e3c76ca4e1dc75577c2f4d9a5e944ad21625f927a7d11064987e154da1
pGx: 54fade0e45f3b8e3ae5301cd67c89ea3e3546c9a221cb9987c804f18fad35ce31fe3253ab6235105cccac4a737e57370224ba63a78519a2fcc8f39f7e8975dc693ff83dd17878e114950e43975b61b537300fc2a595a7234c0ecf99e0442da2ee7b775c234917823cbe9213a7ac41f36473b06f19969a254d222ffedb5eec07b
secret: 15e58c65ce35a90de592d112175e3a529dc627f8544baa7261c13d93bc5ef71d71a5a6b970ba93c9a53917acb7562936f8cb0d2bbec5afca64de6f71bd554641959d44d00e1e740abcf61b0007086ee12078e0deab3454bb91193c8870a9d29f59944deb6a68232a597aa1578100e1c777817dff08423db4b6a24d7921128c41
hash2: fe709b3e807dcaa9a9787f6a2f64475e9a70c1d8
secret2: 36353433323115e58c65ce35a90de592d112175e3a529dc627f8544baa7261c13d93bc5ef71d71a5a6b970ba93c9a53917acb7562936f8cb0d2bbec5afca64de6f71bd554641959d44d00e1e740abcf61b0007086ee12078e0deab3454bb91193c8870a9d29f59944deb6a68232a597aa1578100e1c777817dff08423db4b6a24d7921128c41
hash3: fe709b3e807dcaa9a9787f6a2f64475e9a70c1d8
Pin OK :)

dest_hash: 3af77897f91a456d791e9deebc7acda2012c145f
SKPrime: eeb86e6af43ca9c01403b102812f9b1fd3bed97a
SKPrimeHash: 8b85efe37751a4099f6d9ef14d385f8cc0f8f6f2
Token: '037739871315caef138547b03e348b72', session: '1'
```

**remote_example**

```
$ python remote_example.py
Traceback (most recent call last):
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 135, in read_http_response
    status_code, reason, headers = await read_response(self.reader)
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\http.py", line 130, in read_response
    raise ValueError(f"unsupported HTTP version: {d(version)}")
ValueError: unsupported HTTP version: HTTP/1.0

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "remote_example.py", line 42, in <module>
    loop.run_until_complete(main())
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\asyncio\base_events.py", line 568, in run_until_complete
    return future.result()
  File "remote_example.py", line 32, in main
    await remote.start_listening()
  File "D:\T151602\Desktop\home-assistant-core\homeassistant\components\samsungtv\encrypted\remote.py", line 123, in start_listening
    await self._open()
  File "D:\T151602\Desktop\home-assistant-core\homeassistant\components\samsungtv\encrypted\remote.py", line 113, in _open
    connection = await connect(url, open_timeout=self._timeout)
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 650, in __await_impl_timeout__
    return await asyncio.wait_for(self.__await_impl__(), self.open_timeout)
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\asyncio\tasks.py", line 384, in wait_for
    return await fut
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 663, in __await_impl__
    extra_headers=protocol.extra_headers,
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 322, in handshake
    status_code, response_headers = await self.read_http_response()
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 141, in read_http_response
    raise InvalidMessage("did not receive a valid HTTP response") from exc
websockets.exceptions.InvalidMessage: did not receive a valid HTTP response
```

epenet commented:
It's great that authenticator is working! A big load off my mind... For the remote issue, can you add logging?
```python
import logging
logging.basicConfig(level=logging.DEBUG)
```

sermayoral commented:
@epenet here you are:

```
$ python remote_example.py
DEBUG:asyncio:Using selector: SelectSelector
DEBUG:remote:Tx: GET http://192.168.33.32:8080/socket.io/1/?t=1646996604342
DEBUG:charset_normalizer:Encoding detection: latin_1 is most likely the one.
DEBUG:remote:Rx: <?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
         "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 - Forbidden</title>
 </head>
 <body>
  <h1>403 - Forbidden</h1>
 </body>
</html>

DEBUG:charset_normalizer:Encoding detection: latin_1 is most likely the one.
DEBUG:remote:WS url ws://192.168.33.32:8080/socket.io/1/websocket/<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
         "http
DEBUG:websockets.client:= connection is CONNECTING
DEBUG:websockets.client:> GET /socket.io/1/websocket/<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
         "http HTTP/1.1
DEBUG:websockets.client:> Host: 192.168.33.32:8080
DEBUG:websockets.client:> Upgrade: websocket
DEBUG:websockets.client:> Connection: Upgrade
DEBUG:websockets.client:> Sec-WebSocket-Key: UocjzsFiq8ZkV9OeWpqWyA==
DEBUG:websockets.client:> Sec-WebSocket-Version: 13
DEBUG:websockets.client:> Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
DEBUG:websockets.client:> User-Agent: Python/3.7 websockets/10.2
DEBUG:websockets.client:! failing connection with code 1006
DEBUG:websockets.client:x half-closing TCP connection
DEBUG:websockets.client:= connection is CLOSED
Traceback (most recent call last):
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 135, in read_http_response
    status_code, reason, headers = await read_response(self.reader)
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\http.py", line 130, in read_response
    raise ValueError(f"unsupported HTTP version: {d(version)}")
ValueError: unsupported HTTP version: HTTP/1.0

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "remote_example.py", line 45, in <module>
    loop.run_until_complete(main())
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\asyncio\base_events.py", line 568, in run_until_complete
    return future.result()
  File "remote_example.py", line 35, in main
    await remote.start_listening()
  File "D:\T151602\Desktop\home-assistant-core\homeassistant\components\samsungtv\encrypted\remote.py", line 123, in start_listening
    await self._open()
  File "D:\T151602\Desktop\home-assistant-core\homeassistant\components\samsungtv\encrypted\remote.py", line 113, in _open
    connection = await connect(url, open_timeout=self._timeout)
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 650, in __await_impl_timeout__
    return await asyncio.wait_for(self.__await_impl__(), self.open_timeout)
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\asyncio\tasks.py", line 384, in wait_for
    return await fut
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 663, in __await_impl__
    extra_headers=protocol.extra_headers,
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 322, in handshake
    status_code, response_headers = await self.read_http_response()
  File "D:\T151602\Desarrollo\tools\python-3.7.0\lib\site-packages\websockets\legacy\client.py", line 141, in read_http_response
    raise InvalidMessage("did not receive a valid HTTP response") from exc
websockets.exceptions.InvalidMessage: did not receive a valid HTTP response
```

sermayoral commented:
From now i will use pastebin, sorry.

A question. The token and sessionid received with the original method (get_token.py) should be valid with the remote_example.py script. Doesn't it?

epenet commented:
Do the urls match what is used in the custom component?

epenet commented:
> A question. The token and sessionid received with the original method (get_token.py) should be valid with the remote_example.py script. Doesn't it?

Yes, the token and sessionid received in the previous method (get_token.py) should be compatible.

epenet commented:
Note: I see in your error report that you are using port 8080. I think instead it should be port 8000:
- authenticator uses port 8080
- remote uses port 8000

sermayoral commented:
@epenet Yeah, you are right. port 8000 does the trick

In the test I have changed the KEY_POWER to KEY_POWEROFF to turn off the TV, and it worked. Here is the log: https://pastebin.com/gYyJqgf4

epenet commented:
That's fantastic. Now we just need to make a proper package with:
- `authenticator.py`
- `command.py`
- `exceptions.py`
- `remote.py`
- `session.py`

Dependencies:
- `aiohttp>=3.8.1`
- ~`py3rijndael>=0.3.3`~ no longer needed
- ~`pycryptodome>=3.12.0`~ replaced by `cryptography>=35.0.0`
- `websockets>=10.2`

epenet commented:
Note: I won't be able to work on it this week-end. I'll pick it up again next week.
Feel free to keep testing, or to re-use the code in your custom component (since it is async it will improve the performance).

sermayoral commented:
Thanks @epenet. anything you need, tell me. Have a nice weekend!!

bdraco commented on homeassistant/components/samsungtv/encrypted/authenticator.py:
```diff
@@ -0,0 +1,316 @@
+import struct
+from typing import Any
+
+from Crypto.Cipher import AES
```
Home Assistant uses `cryptography` 

There should be an almost identical interface available without the need to load another crypto lib

epenet replied:
I have adjusted to use cryptography (>= 35.0.0)

epenet commented:
On top of dropping `pycryptodome` dependency, I have also managed to drop the `py3rijndael` dependency

These are the updated dependencies:
- `aiohttp>=3.8.1`
- `cryptography>=35.0.0`
- `websockets>=10.2`

epenet commented:
I have just now pushed some new updates, with an adjusted config-flow.
@ooohaaa you should now be able to test it simply by using the steps described in https://github.com/epenet/home-assistant-core/pull/71#issuecomment-1064488784

If it works as it should:
- it should detect your TV as using the websocket method
- when you press a key it should fail, and automatically fallback to encrypted method + reload
- then you should have a reconfigure option in HomeAssistant, and you should be able to follow the steps to get a token+session_id
- finally it should automatically reload again and you should be able to control your TV

ooohaaa commented:
Looks much better now! 😃
But unfortunately I have now controls. Here is what I have seen:

- When clicking 'reconfigure', Pin was not accepted. Was trying two times. Pin was definitely correct - no typo.
    - [core_log_reconfigure_pin_NOK.txt](https://github.com/epenet/home-assistant-core/files/8248161/core_log_reconfigure_pin_NOK.txt)
    - ![2022-03-14 20_19_06-Configuration – Home Assistant](https://user-images.githubusercontent.com/95161854/158250868-40a4992a-169a-4e55-b7e0-f3aece0e9569.png)

- After switching off/on TV, I tried 'reconfigure' again. TV didn't show a pin.
- Restarted HA Host. 'Reconfigure' and entering pin was successful now.
- Unfortunately there are no controls
    - [core_log_reconfigure_pin_successful.txt](https://github.com/epenet/home-assistant-core/files/8248175/core_log_reconfigure_pin_successful.txt)
    - ![2022-03-14 20_44_10-Configuration – Home Assistant](https://user-images.githubusercontent.com/95161854/158251019-4f00b2c8-fe5d-4465-ac7e-3d18b8df3b87.png)


Let me know if I can test/try anything!

epenet commented:
Would you be able to add the new diagnostics please?

ooohaaa commented:
Here it is: [config_entry-samsungtv-b99274a99c10b742928ce7905fe0f88b.json.txt](https://github.com/epenet/home-assistant-core/files/8248543/config_entry-samsungtv-b99274a99c10b742928ce7905fe0f88b.json.txt)

epenet commented:
Aha the port is wrong. It stayed as 8002.
Ill check tomorrow

epenet commented:
> Aha the port is wrong. It stayed as 8002. Ill check tomorrow

The latest code should automatically adjust the port from 8002 to 8000 for encrypted connections.

ooohaaa commented:
I gave it a try. I can see the controls now, but the TV does not react when clicking the controls.
Here is the log and diagnostics:
[core_log.txt](https://github.com/epenet/home-assistant-core/files/8256948/core_log.txt)
[config_entry-samsungtv-d7e57c3da56a398f02ed492e44c1fe9d.json.txt](https://github.com/epenet/home-assistant-core/files/8256951/config_entry-samsungtv-d7e57c3da56a398f02ed492e44c1fe9d.json.txt)

epenet commented:
You log seems to have had an issue with the migration from websocket to legacy.
Did you try to restart the integration, or restart Home Assistant after the migration?

```console
2022-03-15 21:32:27 ERROR (MainThread) [homeassistant.components.media_player] Platform samsungtv does not generate unique IDs. ID 223da676-497a-4e06-9507-5e27ec4f0fb3 already exists - ignoring media_player.tv_ue48ju6470
2022-03-15 21:32:33 WARNING (MainThread) [homeassistant.helpers.service] Unable to find referenced entities media_player.tv_ue48ju6470
2022-03-15 21:32:40 WARNING (MainThread) [homeassistant.helpers.service] Unable to find referenced entities media_player.tv_ue48ju6470
```

ooohaaa commented:
Sorry, I don't know what is meant by "migration from websocket to legacy". It's also unclear what is meant with "restart HA after migration".

However, I retried it again today to document the steps I did...

1. Updated Core&Host. core-2022.3.5 / supervisor-2022.03.4
2. Delete Integration
3. Delete old 'samsungtv' folder inside custom_componnets (ftp)
4. Copy new 'samsungtv' folder to custom_components (ftp)
5. Add version tag `"version": "2022.4.0-20220316",` in manifest
6. Restart Host
7. TV was not detected.
8. Restart Host
9. TV still not detected. Here is the log. [core_log - TV not detected after Restart.txt](https://github.com/epenet/home-assistant-core/files/8273370/core_log.-.TV.not.detected.after.Restart.txt)

So currently I have a different failure. Can't say if this is related to the core&supervisor update.

Any Ideas?

epenet commented:
The new library is available, but it seems I missed a refactor and something is still pointing at the old deleted encrypted files.
I'll adjust in the morning (CET)

epenet commented:
I think this is now ready to go. I have opened the core PR: https://github.com/home-assistant/core/pull/68301

Please test "in anger" now :)

ooohaaa commented:
Unfortunately it's not working yet. Tests werde done with commit [7d5ba57](https://github.com/epenet/home-assistant-core/pull/71/commits/7d5ba57e21fa436ff2a299c4c6449b9b02bd97c9)

Here are my logs/diagnostics:

[config_entry-samsungtv-83cfbb48e25194ba1d341394ccd552ef.json(1).txt](https://github.com/epenet/home-assistant-core/files/8291542/config_entry-samsungtv-83cfbb48e25194ba1d341394ccd552ef.json.1.txt)
[core_log_2022-03-17.txt](https://github.com/epenet/home-assistant-core/files/8291550/core_log_2022-03-17.txt)

epenet commented:
@sermayoral have you had a chance to make some more tests?

sermayoral commented:
> @sermayoral have you had a chance to make some more tests?

Yeah. What do you need?

epenet commented:
@sermayoral
When @bdraco tests `ha-samsungtv-encrypted` with his TV (UN75JU641D), he is able to control the volume (but not power off). When he tests with this PR he is unable to control the volume.

So I am trying to understand where the problem might come from.

Can you please check how well this code works on your machine?
https://github.com/xchwarze/samsung-tv-ws-api/blob/master/example/encrypted_remote.py

Maybe double try all the commands, and check that they work?
`KEY_MUTE`, `KEY_VOLUP`, `KEY_VOLDOWN`, `KEY_PLAY`, `KEY_PAUSE`, `KEY_FF`, `KEY_REWIND`

Maybe try to chain them, to ensure that they work when the remote stays open?
```python
        await remote.start_listening()

        await remote.send_command(SendRemoteKey.click("KEY_MUTE"))

        await asyncio.sleep(15)

        await remote.send_command(SendRemoteKey.click("KEY_MUTE"))

        await asyncio.sleep(15)

        await remote.send_command(SendRemoteKey.click("KEY_VOLDOWN"))

        await asyncio.sleep(15)

        await remote.send_command(SendRemoteKey.click("KEY_VOLUP"))

        await remote.close()
```

sermayoral commented:
@epenet here you are. I have the UE50H6400 model

### Separated

- KEY_POWEROFF: Success

https://pastebin.com/KWZmcXs6

Note: It doesn't work in the YouTube app. The TV doesn't turn off when a YouTube video is playing (it also happen in my custom component...). Walkaround is to turn off using HDMI-CEC, or changing to another source (for example HDMI1 or TV) before sending the KEY_POWEROFF command...

- KEY_MUTE: Success

https://pastebin.com/CG9L2REj

- KEY_VOLUP: Success

https://pastebin.com/VYE0GJd3

- KEY_VOLDOWN: Success

https://pastebin.com/eW85U6Yx

- KEY_PAUSE: Success

https://pastebin.com/1vUs15iT

- KEY_PLAY: Success

https://pastebin.com/XqJv5A5Q

- KEY_FF: Success

https://pastebin.com/YZxTjUd1

- KEY_REWIND: Success

https://pastebin.com/313HuuYv

### Chained

Only the first two commands work. I think the connection last around 20-25 seconds, so the following commands fail:

https://pastebin.com/xc7uwb6w

With the following code, it works:

```
        await remote.start_listening()
        await remote.send_command(SendRemoteKey.click("KEY_MUTE"))
        await asyncio.sleep(3)
        await remote.send_command(SendRemoteKey.click("KEY_MUTE"))
        await asyncio.sleep(3)
        await remote.send_command(SendRemoteKey.click("KEY_VOLDOWN"))
        await asyncio.sleep(3)
        await remote.send_command(SendRemoteKey.click("KEY_VOLUP"))
        await remote.close()
```

https://pastebin.com/CikbBdAK

I hope it helps :-)

epenet commented:
Thanks, it's great to know that the connection doesn't stay up regardless.
I'll make some further adjustments to the component here.

epenet commented:
@sermayoral can you please try now revision 3f5ff36c66ee614c89d3417f43ea2cea11ec5314 inside Home Assistant?
- create a new custom_components\samsungtv folder in your HA config folder
- copy all the files (including the translations) into the new folder. They can be downloaded from this link: https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/epenet/home-assistant-core/tree/3f5ff36c66ee614c89d3417f43ea2cea11ec5314/homeassistant/components/samsungtv
- edit custom_components\samsungtv\manifest.json to add a new version key

I have made so many commits that I think it's better to use the commit id for tracking.

epenet commented:
@sermayoral did you get a chance to test inside HomeAssistant?

sermayoral commented:
@epenet sorry, I have had no time until now. 

The TV asked me for a PIN, I introduced the PIN and the TV was successfully added. The power status is working well, but commands are not working. Poweroff, volume up, down, etc are not working :-(

Here is the log: https://pastebin.com/piPM2xAs

epenet commented:
@sermayoral thanks for confirming.
So the problem is inside the PR, not inside the upstream library.

It's driving me crazy! What is different between the PR and the direct connection?

Can you please add extra logging, ready for the next tests?
```yaml
logger:
    default: warning
    logs:
        homeassistant.components.samsungtv: debug
        custom_components.samsungtv: debug
        samsungtvws: debug
        samsungctl: debug
```

sermayoral commented:
@epenet  I have certain suspicions that the token and sessionid obtained with your library is not valid. The last tests i have done with your library were with the token and sessionId obtained with the get_token.py from my custom_component (those configured in HA). I used the data I already had so I didn't have to re-pair the TV (everytime the TV shows a PIN, the previous token become invalid).

Late yesterday, I did a basic test of getting a token and running a command again with your library, and it didn't work. I added your library token to my custom component and it didn't work either. Maybe the problem is there

epenet commented:
@sermayoral can you adjust the session_id and token inside you `config\.storage\core.config_entries` file?
If you put in there the same session_id and token as you are using in your custom component, does it work?

epenet commented:
I have traced the issue to `num_rounds` in `py3rijndael`.

I wrongly assumed that the `py3rijndael` public library matched the folder in `PySmartCrypto`, but that is not the case!

Public library has: https://github.com/meyt/py3rijndael/blob/3c2a6a3776b38b1f988b3c9148b3bca3c89f3112/py3rijndael/constants.py#L7-L11
```python
num_rounds = {
    16: {16: 10, 24: 12, 32: 14},
    24: {16: 12, 24: 12, 32: 14},
    32: {16: 14, 24: 14, 32: 14}
}
```

PySmartCrypto has: https://github.com/sermayoral/ha-samsungtv-encrypted/blob/da0f13fe3bcffc834e84fb0d46b8d9066bb2681e/custom_components/samsungtv_encrypted/PySmartCrypto/py3rijndael/constants.py#L7-L9
```python
num_rounds = {
    16: {16: 3}
}
```

Now that I know the root cause, I need to decide how best to handle that in `samsungtvws`.

sermayoral commented:
@epenet great news!! Sorry, I can't test it until night CET (i'm at work right now).

Tell me what do you need to test after 3-4 hours, and i will test in the night

epenet commented:
@sermayoral the config flow just stores the session_id and token inside `config\.storage\core.config_entries` file
So if you replace the two values manually with values that you know are working from your custom component, and then restart Home Assistant, it will confirm that the issue is only inside the `authenticator` class

sermayoral commented:
@epenet ok. You need me to test the same thing you said me before. Well, i'll do this night, as soon as possible.

Thanks!

epenet commented:
New commit 9e820452dcbadd8dc24a39cdcb9a85fba7eed5e4 available. It doesn't have the new authenticator available yet, but it has been rebased on latest dev branch :

- create a new custom_components\samsungtv folder in your HA config folder
- copy all the files (including the translations) into the new folder. They can be downloaded from this link: https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/epenet/home-assistant-core/tree/9e820452dcbadd8dc24a39cdcb9a85fba7eed5e4/homeassistant/components/samsungtv
- edit custom_components\samsungtv\manifest.json to add a new version key

epenet commented:
@sermayoral

Separately, if you have a chance, it would be nice if you could try to generate a new token/session_id with the latest version of the library, now that https://github.com/xchwarze/samsung-tv-ws-api/pull/104 is merged.
`pip3 install "git+https://github.com/xchwarze/samsung-tv-ws-api.git#egg=samsungtvws[async,encrypted]"`

Then you can try the new token/session_id inside your custom component before we build a new release.

sermayoral commented:
> @sermayoral the config flow just stores the session_id and token inside `config\.storage\core.config_entries` file So if you replace the two values manually with values that you know are working from your custom component, and then restart Home Assistant, it will confirm that the issue is only inside the `authenticator` class

I can confirm, replacing the token and session_id got from get_token.py (custom_component,), its .......

Edit: ............. (drums)

sermayoral commented:
¡WORKING!

Here is the log: https://pastebin.com/TjDjSZMj

epenet commented:
Great! It was driving me insane!
Now just need confirmation that https://github.com/xchwarze/samsung-tv-ws-api/pull/104 fixes the authenticator

sermayoral commented:
@epenet can be tested using encrypted_authenticator and encrypted_remote.py of master branch?

epenet commented:
Yes master branch of https://github.com/xchwarze/samsung-tv-ws-api

sermayoral commented:
> Yes master branch of https://github.com/xchwarze/samsung-tv-ws-api

It doesn't work 😥

epenet commented:
The token and session id from the new authenticator don't work in your custom component?
Argh!
Do you have the authenticator logs?
I will run the http traces in my unit test.
:(

sermayoral commented:
> The token and session id from the new authenticator don't work in your custom component? Argh! Do you have the authenticator logs? I will run the http traces in my unit test. :(

It isn't exact. Token and sessionId got from encrypted_authenticator doesn't work on encrypted_remote.py

sermayoral commented:
I'm using

```
commit 8914a900630fb03ee088e77ccd9931a02438525a (HEAD -> master, origin/master, origin/HEAD)
Merge: 3b8bcf1 5143669
Author: DSR! <xchwarze@gmail.com>
Date:   Tue Mar 22 12:58:50 2022 -0300

    Merge pull request #104 from epenet/fix-authenticator

    Fix issue with authenticator token
```

Token and Session generated from samsung-tv-ws-api/example/encrypted_authenticator.py doesn't work on samsung-tv-ws-api/example/encrypted_remote.py

Logs from authenticator: https://pastebin.com/qDDQ0KHP
Logs from remote: https://pastebin.com/zJVEuK4i

Token and Session generated from get_token.py (custom) and tested on samsung-tv-ws-api/example/encrypted_remote.py. It works.

Logs from remote: https://pastebin.com/K4cXKzBp

epenet commented:
I have double-check the traces. It seems that you are not using the latest code.

If I use commit 3b8bcf153a23eedde0803ca14c50ae5e89197516, then I get your token `6d0ec7984bcd8e26e85508b7faa2bcff`
If I use commit 8914a900630fb03ee088e77ccd9931a02438525a, then I get a different token `ba6e36b389ce796a539b7ebc54710699`

I think token `ba6e36b389ce796a539b7ebc54710699` would work on your setup

epenet commented:
Did you try to install samsungtvws like this?
`pip3 install "git+https://github.com/xchwarze/samsung-tv-ws-api.git#egg=samsungtvws[async,encrypted]"`

sermayoral commented:
> I have double-check the traces. It seems that you are not using the latest code.
> 
> If I use commit 3b8bcf153a23eedde0803ca14c50ae5e89197516, then I get your token `6d0ec7984bcd8e26e85508b7faa2bcff` If I use commit 8914a900630fb03ee088e77ccd9931a02438525a, then I get a different token `ba6e36b389ce796a539b7ebc54710699`
> 
> I think token `ba6e36b389ce796a539b7ebc54710699` would work on your setup

How have you got a token? Have you mocked the PIN?

epenet commented:
The PIN, token, sessionid and all http responses appear in the log, so I can replicate the flow 100%.
This is why I think you didn't pick up latest code.

epenet commented:
Old code = same token as log file
New code = different token `ba6e36b389ce796a539b7ebc54710699`

sermayoral commented:
@epenet sorry, maybe i'm stupid or something like that, but it doesn't work. Here is the full process log:

https://pastebin.com/BQwWjMBY

epenet commented:
> @epenet sorry, maybe i'm stupid or something like that, but it doesn't work. Here is the full process log:

I don't think you're stupid 🤣and I don't understand either.

Your log shows PIN '6423', token: 'fe26869dcc5569bf18a5229810e83657', session: '1'
And when I run it here on latest master, I get PIN '6423', token: '634fa675faf6e8a8e5140cdc1b80039e', session: '1'

So there is a bad reference somewhere. I just don't know where.

Anyway, if you try "my" new token `634fa675faf6e8a8e5140cdc1b80039e` does it work on your TV?

epenet commented:
If my token works, then we can ask @xchwarze to publish the updated package.

sermayoral commented:
Sorry, i have been testing and the PIN '6423' is already invalidated.

Can you generate me a new one with the PIN '0907' ??

Edit: I got Token: 'a727b6e8bd9e719d636d8148d3b817bf', session: '1' that, of course, it doesnt work :-(

epenet commented:
I cannot generate "from PIN" I can only generate "from log file"
You run trace, with correct PIN and I give you the correct token and you try my token in remote

sermayoral commented:
Here you are: https://pastebin.com/349G8w6n

epenet commented:
> Here you are: https://pastebin.com/349G8w6n

Correct token should be `e0d8db0463705050041b239223400ca4`. If you put that inside remote test it should work.

sermayoral commented:
@epenet it works!!

https://pastebin.com/TVRkzzGi

epenet commented:
The upstream library has been updated, so it should now all work correctly inside Home Assistant as a custom component.

I have pushed commit 2a596204284a666c43038fe712c7a5a71b37e38e if anyone wants to have a go:
https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/epenet/home-assistant-core/tree/2a596204284a666c43038fe712c7a5a71b37e38e/homeassistant/components/samsungtv

@ooohaaa you should be able to test it as well - no need for Python knowledge hopefully.

sermayoral commented:
KEY_POWER not working in all encrypted models, like mine

bdraco commented:
KEY_POWER is working for me but it only works once and then I have to resetup

epenet commented on homeassistant/components/samsungtv/bridge.py:
```diff
@@ -737,9 +763,9 @@ async def _async_get_remote_under_lock(
-        raise HomeAssistantError(
-            "Sending commands to encrypted TVs is not yet supported"
-        )
+        await self._async_send_commands([SendEncryptedRemoteKey.click("KEY_POWER")])
```
Maybe we should send both?
```suggestion
        # some models use KEY_POWEROFF (eg. UE50H6400)
        await self._async_send_commands([SendEncryptedRemoteKey.click("KEY_POWEROFF")])
        # some models use KEY_POWER (eg. UN75JU641D)
        await asyncio.sleep(1)
        await self._async_send_commands([SendEncryptedRemoteKey.click("KEY_POWER")])
```

bdraco replied:
Does KEY_POWEROFF work on any encrypted models?

bdraco replied:
If its confirmed to work on `UE50H6400` then probably sending both makes sense

bdraco replied:
Or maybe if we know its an `H` send `KEY_POWEROFF` and a `J` send `KEY_POWER`

epenet replied:
If we want to support more models, I would say send both.
If we want to be strict, then maybe default to `KEY_POWEROFF` and use a known model list to override with `POWER_KEY`

bdraco replied:
If you don't know for sure, send both.  If you do know for sure which one it takes, send that one

epenet replied:
How about this?
```suggestion
        power_off_commands: list[SendEncryptedRemoteKey] = []
        if self._model in self._MODEL_USES_POWER_OFF:
            power_off_commands.append(SendEncryptedRemoteKey.click("KEY_POWEROFF"))
        elif self._model in self._MODEL_USES_POWER:
            power_off_commands.append(SendEncryptedRemoteKey.click("KEY_POWER"))
        else:
            LOGGER.info(
                "Unknown power_off command for %s (%s): sending KEY_POWEROFF and KEY_POWER",
                self.host,
                self._model,
            )
            power_off_commands.append(SendEncryptedRemoteKey.click("KEY_POWEROFF"))
            power_off_commands.append(SendEncryptedRemoteKey.click("KEY_POWER"))
        await self._async_send_commands(power_off_commands)

```

With corresponding class definition
```python
class SamsungTVEncryptedBridge(SamsungTVBridge):
    """The Bridge for Encrypted WebSocket TVs (J/H models)."""

    _MODEL_USES_POWER_OFF = set("UE50H6400")
    _MODEL_USES_POWER = set("UN75JU641D")

    def __init__(...) -> None:
        """Initialize Bridge."""
        self._model: str | None = None
        if entry_data:
            self._model = entry_data.get(CONF_MODEL)
```

sermayoral replied:
In my case, KEY_POWEROFF always works except when the YouTube app is open. In that case KEY_POWEROFF acts as a RETURN command, but the TV doesn't turn off. I guess that's a misunderstanding of the YouTube app.

Sending both commands to turn off the TV may be the best and easy solution, but I would also enable the 'on_turn_off' property so that, if the user defines it, a more reliable method of turning it on and off is enabled, such as using a relay, IR or HDMI-CEC (same philosophy as the 'on_turn_on' property)

epenet replied:
For now, I have purposefully ignored `on_turn_on` / `on_turn_off` properties, at least until we decide what we do about the old `on_turn_on` method. See https://github.com/home-assistant/core/pull/68185

epenet replied:
I have pushed up my changes for the double-key `KEY_POWER`/`KEY_POWEROFF`.
I have used `frame.report` to log it when unknown, so it should only get logged once and encourage users to add their model.

epenet replied:
@ooohaaa it would be interesting to have your results on model `UE48JU6400`.
Does it work with `KEY_POWER`? or with `KEY_POWEROFF`? or neither?
Does the token still work after power off, or does it only work on first attempt?

ooohaaa replied:
@epenet I will be able to test this weekend. Starting with next monday I won't be available for two weeks.

> Does it work with `KEY_POWER`? or with `KEY_POWEROFF`? or neither?

How can I test this?
Or shall I just run the new custom componet like described below?

> The upstream library has been updated, so it should now all work correctly inside Home Assistant as a custom component.
> 
> I have pushed commit [2a59620](https://github.com/epenet/home-assistant-core/commit/2a596204284a666c43038fe712c7a5a71b37e38e) if anyone wants to have a go: https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/epenet/home-assistant-core/tree/2a596204284a666c43038fe712c7a5a71b37e38e/homeassistant/components/samsungtv
> 
> @ooohaaa you should be able to test it as well - no need for Python knowledge hopefully.

epenet replied:
You can use the new commit: a4ab29647f653142613e8cc52430a42f749ecb5e:
https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/epenet/home-assistant-core/tree/a4ab29647f653142613e8cc52430a42f749ecb5e/homeassistant/components/samsungtv

By default, it will send both commands and should report a warning in the logs.
Then to try just one command you can adjust lines 615-616 in `bridge.py` and see which one works for you:
https://github.com/epenet/home-assistant-core/blob/a4ab29647f653142613e8cc52430a42f749ecb5e/homeassistant/components/samsungtv/bridge.py#L615-L616

sermayoral replied:
@epenet i'm agreed with your solution, but there is an important thing you must know. The Samsung model identifier is composed of several parts:

- First two characters means region/continent (UE is from Europe)
- Following two characters means screen size (50 means 50")
- The rest of characters are the real model, regardless of screen size or region. In my case H6400

I mean, an UN45H6400 model will work the same way as UE50H6400.

epenet replied:
😆 yes I already adjusted that in https://github.com/epenet/home-assistant-core/pull/71/commits/63e96ec6d98ef8b2991608eba6005b9d05f0d642

epenet commented:
@sermayoral if you could make one last test with b77b1f76fe9ab37fcf74c4497adf54e1a16898b6 I think we can merge it.
Please add a note on https://github.com/home-assistant/core/pull/68301 once you've had a chance to test it one last time.

epenet commented:
@ooohaaa we believe that JU6400 models react the same way as JU641D models, but if you could confirm that would be great.
You too can test with commit b77b1f76fe9ab37fcf74c4497adf54e1a16898b6

It would be interesting to know if your token stays working after power-off, or if like @bdraco it becomes invalid

sermayoral commented:
> @ooohaaa we believe that JU6400 models react the same way as JU641D models, but if you could confirm that would be great. You too can test with commit [b77b1f7](https://github.com/epenet/home-assistant-core/commit/b77b1f76fe9ab37fcf74c4497adf54e1a16898b6)
> 
> It would be interesting to know if your token stays working after power-off, or if like @bdraco it becomes invalid

My token doesn't expire. It only expires when you start a pair process again (when TV shows a new PIN)

epenet commented:
> @ooohaaa we believe that JU6400 models react the same way as JU641D models,

epenet commented:
> > @ooohaaa we believe that JU6400 models react the same way as JU641D models, but if you could confirm that would be great. You too can test with commit [b77b1f7](https://github.com/epenet/home-assistant-core/commit/b77b1f76fe9ab37fcf74c4497adf54e1a16898b6)
> > It would be interesting to know if your token stays working after power-off, or if like @bdraco it becomes invalid
> 
> My token doesn't expire. It only expires when you start a pair process again (when TV shows a new PIN)

Yes - but we were wondering if it expired for @ooohaaa on JU6400 model.

sermayoral commented:
> @sermayoral if you could make one last test with [b77b1f7](https://github.com/epenet/home-assistant-core/commit/b77b1f76fe9ab37fcf74c4497adf54e1a16898b6) I think we can merge it. Please add a note on [home-assistant#68301](https://github.com/home-assistant/core/pull/68301) once you've had a chance to test it one last time.

Working perfectly. Great work!!! Here is the log: https://pastebin.com/MT3DYJT9

A minor thing is the windows that ask you for a pin shows in that way:

![image](https://user-images.githubusercontent.com/30116095/159966322-fdd8f46d-66d7-4808-b4ed-80b8b7caebb1.png)

I think this windows showed a text asking for a PIN before. Maybe is a problem with the explorer caché.

epenet commented:
https://github.com/home-assistant/core/pull/68301 has been merged so I am closing it here.
New issues should be logged there.

ooohaaa commented:
@epenet I have tested with this download link :https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/epenet/home-assistant-core/tree/a4ab29647f653142613e8cc52430a42f749ecb5e/homeassistant/components/samsungtv

I was able to Mute, Volume up/down and Power off. 🎉

Unfortunately Play/Pause Button does not work as expected. Seems like only the  "Play" command is sent - but never the "Pause" command. Here is what I have seen:
- If I press pause on the original remote, and Play/Pause in HA afterwards, the TV starts playing again.
- If TV is already playing and I click Play/Pause in HA the TV shows some channe info - its the same behaviour when I press Play on the orgiginal remote when TV is already playing.


[ha_core_log.txt](https://github.com/epenet/home-assistant-core/files/8356399/ha_core_log.txt)
For some reason. downloading the diagnostics does not work. 🤷‍♂️


> @ooohaaa we believe that JU6400 models react the same way as JU641D models, but if you could confirm that would be great. You too can test with commit [b77b1f7](https://github.com/epenet/home-assistant-core/commit/b77b1f76fe9ab37fcf74c4497adf54e1a16898b6)
> 
> It would be interesting to know if your token stays working after power-off, or if like @bdraco it becomes invalid

As written above I have tested with older version. I can repeat the tests if you send a new download url.
The integration worked after switching on/off (with remote), but not after an power reset.

epenet commented:
@ooohaaa everything has been merged into core now, so it will be available as part of 2022.04

If you want to do some additional testing before go-live, this is the new download link: https://minhaskamal.github.io/DownGit/#/home?url=https://github.com/home-assistant/core/tree/dev/homeassistant/components/samsungtv
